xonsh/tools.py

```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
         """A lazy value getter."""
         return self._cmds_cache.get(key, default)
 
+    def locate_binary(self, name):
+        """Locates an executable on the file system."""
+
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+
+            local_bins = [full_name for full_name in self.get_possible_names(name)
+                          if os.path.isfile(full_name)]
+            if local_bins:
+                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self.all_commands.keys()
+
+        if matches:
+            return self.all_commands[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
+    def lazy_locate_binary(self, name):
+        """Locates an executable in the cache."""
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self._cmds_cache.keys()
+
+        if matches:
+            return self._cmds_cache[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
```

Comment:
I find them hugely distracting and harder to understand algorithms because less code fits on a page :). Also, PEP8 basically says to leave them out.  I think that this is mostly a style thing, so I defer to PEP8